src/m3ninx/doc/types.go

```diff
@@ -129,4 +129,17 @@ type OnIndexSeries interface {
 	// The range is inclusive. Note that there may be uncovered gaps within the range.
 	// Returns (0, 0) for an empty range.
 	IndexedRange() (xtime.UnixNano, xtime.UnixNano)
+
+	// ReconciledOnIndexSeries attempts to retrieve the most recent index entry from the
+	// shard if the entry this method was called on was never inserted there. If there
+	// is an error during retrieval, simply returns the current entry. Additionally,
+	// returns a cleanup function to run once finished using the reconciled entry and
+	// a boolean value indicating whether the result came from reconciliation or not.
+	// Cleanup function must be called once done with the reconciled entry so that
+	// reader and writer counts are accurately updated.
+	ReconciledOnIndexSeries() (OnIndexSeries, ReconciledOnIndexSeriesCleanupFn, bool)
 }
+
+// ReconciledOnIndexSeriesCleanupFn is a function for performing cleanup when
+// ReconciledOnIndexSeries is called.
+type ReconciledOnIndexSeriesCleanupFn func()
```

Comment:
nit: This could just be `CleanupFn` given it could be used for other returns/etc since it's just a `func()`? 

To be even "cleaner" you could return a `resource.SimpleCloser`.
https://pkg.go.dev/github.com/m3db/m3@v1.3.0/src/x/resource#SimpleCloser

e.g.
```go
import (
  "github.com/m3db/m3/src/x/resource"
)

// then definition
ReconciledOnIndexSeries() (OnIndexSeries, resource.SimpleCloser, bool)

// then when returning e.g. cast the func() to a resource.SimplerCloserFunc which implements resource.SimpleCloser
	return e, resource.SimpleCloserFn(func() {
		e.DecrementReaderWriterCount()
	}), true

// then using the results
entry, closer, ok := entry.ReconciledOnIndexSeries()
defer closer.Close()

```

Reply:
Ah, I like this